### ISSUE TITLE
A Web Worker's type should be made generic

### ISSUE THREAD

**Issue by RebeccaStevens** [microsoft/TypeScript#48396]
# Suggestion

## 🔍 Search Terms

List of keywords you searched for before creating this issue. Write them down here so that others can find this suggestion more easily and help provide feedback.

- webworker
- generic

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Generics should be able to be passed to the `Worker` interface defined in `lib.dom.d.ts`.

It should look something like this:

```ts
export interface Worker<Receive = any, Send = any, SendError = Send> extends EventTarget, AbstractWorker {
    onmessage: ((this: Worker, ev: MessageEvent<Send>) => any) | null;
    onmessageerror: ((this: Worker, ev: MessageEvent<SendError>) => any) | null;
    postMessage(message: Receive, transfer: Transferable[]): void;
    postMessage(message: Receive, options?: StructuredSerializeOptions): void;
	// ...
}
```

## 📃 Motivating Example

It would be nice to be able to type what data is be sent to and from an work.

## 💻 Use Cases

Writing typed web workers.

**Comment by fatcerberus** [comment]
This wouldn’t give you any meaningful type safety though?  It would be similar to making `JSON.parse` generic only to write `JSON.parse<Foo>(str)`—with no compile-time verification that the thing you actually get back is a `Foo`.

**Comment by RebeccaStevens** [comment]
If only you have a reference to the worker, only you can post messages to it and receive messages from it. Thus you know all the types can be send to and from the worker. There aren't going to be any unexpected types.

**Comment by RebeccaStevens** [comment]
@fatcerberus

I don't see this being like `JSON.parse` at all.
If you you create a worker, you know all of the types that can be posted to it, and you know all of the types it can post back.

It seems more like `Array<Foo>` to me.
Sure you can end up with a non-`Foo` value in that array if you cast the array to `any` and push a non-`Foo` value, but `Array<Foo>` is still consider type safe. `Worker`s just seem like the same sort of thing.

**Comment by Pyrolistical** [comment]
Agreed. It seems strange that `MessageEvent` is generic, thus `onmessage` can be typed, but `postMessage` is not.

**Comment by adityamatt** [comment]
Following up on this again, is this being prioritised ?

**Comment by ksbrar** [comment]
+1 for this FR
